File: AppCpp/carpark_queueing_cpp.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <random>
#include <chrono>
#include <iomanip>

using namespace std;

// int percentage(int percent, const vector<int>& list) {
//     for (size_t index = 0; index < list.size(); ++index) {
//         if (list[index] >= percent) {
//             return index;
//         }
//     }
//     return -1;
// }

void modelrun(int arrivalrate, int servicetime, int spaces) {
    // Initialize Variables
    auto start_time = chrono::high_resolution_clock::now();
    int count_arrivals = 0;
    int cyclecount = 10000;
    int carsqueued = 0;
    int queue = 0;
    int queuetime = 0;
    int queuetest = 0;
    double hours = 0.0;

    vector<int> count_carsparked_q(arrivalrate * (servicetime / 600) + 200, 0);
    vector<int> count_carsqueued(arrivalrate * (servicetime / 600) + 200, 0);
    vector<int> carsparked_q;

    vector<int> percentiles = {10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 98, 99};
    mt19937 rng(random_device{}());
    uniform_int_distribution<int> distribution(1, 3600);

    // Simulation Loop
    for (int i = 1; i <= cyclecount * 3600; ++i) {
        if (i > 3600 * 100 && i % 36000 == 0) {
            if (queuetest == static_cast<int>(round(static_cast<double>(carsqueued) / count_arrivals * 10000000)) || i == 3000 * 3600) {
                hours = i / 3600.0;
                break;
            } else {
                queuetest = static_cast<int>(round(static_cast<double>(carsqueued) / count_arrivals * 10000000));
            }
        }

        // Count current carpark utilization
        count_carsparked_q[max(static_cast<int>(carsparked_q.size()) - 1, 0)]++;
        count_carsqueued[queue]++;
        queuetime += queue;

        // Reduce parked cars' time
        if (!carsparked_q.empty()) {
            for (auto& time : carsparked_q) {
                time -= 1;
            }
            if (carsparked_q[0] == 0) {
                carsparked_q.erase(carsparked_q.begin());
                if (queue > 0) {
                    carsparked_q.push_back(servicetime);
                    queue--;
                }
            }
        }

        // Handle arrivals
        if (distribution(rng) <= arrivalrate) {
            count_arrivals++;
            if (static_cast<int>(carsparked_q.size()) - 1 < spaces) {
                carsparked_q.push_back(servicetime);
            } else {
                queue++;
                carsqueued++;
            }
        }
    }

    // Data Outputs
    auto end_time = chrono::high_resolution_clock::now();
    chrono::duration<double> elapsed_time = end_time - start_time;

    cout << "Model completed in " << round(elapsed_time.count()) << " seconds" << endl;
    if (hours == 3000) {
        cout << "0.00001 stability not found. " << hours << " hours of survey data generated.\n";
    } else {
        cout << "Stable solution found after " << hours << " hours of survey data.\n";
    }

    if (cyclecount > 0) {
        cout << "Random Arrivals = " << fixed << setprecision(1)
             << static_cast<double>(count_arrivals) / cyclecount << "\n";
        cout << "Service Time = " << servicetime << "\n";
        cout << "Cars Queued = " << round(static_cast<double>(carsqueued) * 100 / count_arrivals) << "%\n";
        cout << "Average Queue time per Arrival = " << round(static_cast<double>(queuetime) / count_arrivals) << " seconds\n";
    }
}

int main() {
    int arrivalrate = 1000;
    int servicetime = 1000;
    int spaces = 300;

    modelrun(arrivalrate, servicetime, spaces);

    return 0;
}

File: AppCpp/carpark_queueing_cpp.py

#Import Functions Required
import random
import itertools
import time
import sys

arrivalrate = 100
servicetime = 100
spaces = 5
precision = 1

#Define Functions
def percentageoftime(percent,list):
    for index, number in enumerate(list):
        if number >= percent:
            return index

def modelrun(arrivalrate,servicetime,spaces):

    #Intialise All Variables
    start_time = time.time()
    count_arrivals = 0
    count_carsparked_q = [0] * int(arrivalrate * int(servicetime/600)+200)
    count_carsqueued = [0] * int(arrivalrate * int(servicetime/600)+200)
    carsparked_q = []
    cyclecount = int(10000)
    arrival = 0
    carsqueued = 0
    queue = 0
    percentiles = [10,20,30,40,50,60,70,80,90,95,98,99]
    queuetime = 0
    queuetest = 0
    hours = 0

    #Generate Arrivals
    
    for i in range (1,cyclecount * 3600 ):
        if i  > 3600*100 and i%36000 ==0:
            if queuetest == round(carsqueued/count_arrivals,5) or i == 3000*3600:
                hours = i / 3600
                break
            else:
                queuetest = round(carsqueued/count_arrivals,5)
                
                
        # Count current carpark utilisation
        count_carsparked_q[max(len(carsparked_q)-1,0)] += 1
        count_carsqueued[queue] += 1
        queuetime += queue

        # Reduce parked cars time remaining by passing time
        if len(carsparked_q) > 0:
            carsparked_q = [item - 1 for item in carsparked_q]

            # move finished cars out and queued cars in    
            if carsparked_q[0] == 0:
                del carsparked_q[0]
                if queue > 0:
                    carsparked_q.append(servicetime)
                    queue -= 1

        # Check if new car arrived and at to carpark
        arrival = random.randint(1,3600 * precision)
        if arrival <= arrivalrate * precision:
            count_arrivals += 1
            if len(carsparked_q) - 1 < spaces:
                carsparked_q.append(servicetime)
            if len(carsparked_q) - 1 == spaces:
                queue +=1
                carsqueued +=1
            if len(carsparked_q) - 1 > spaces:
                print("more cars than spaces somehow")

    cyclecount = hours

    #Convert Counts to Percentiles
    count_carsqueued = list(itertools.accumulate(count_carsqueued))
    count_carsparked_q = list(itertools.accumulate(count_carsparked_q))
    if cyclecount > 0:
        count_carsqueued = [round(100 * item / (cyclecount * 3600 * precision),2) for item in count_carsqueued]
        count_carsparked_q = [round(100 * item / (cyclecount * 3600 * precision),2) for item in count_carsparked_q]

    for i in range(0,100):
        for index, number in enumerate(count_carsqueued):
                if number >= i:
                    count_carsqueued.append(index)
                    break
    for i in range(0,100):
        for index, number in enumerate(count_carsparked_q):
                if number >= i:
                    if number <= count_carsqueued[0]:
                        count_carsparked_q.append(index)
                        break
                        # Simple Data Outputs
    end_time = time.time()
    elapsed_time = end_time - start_time
    print("Model completed in ", int(round(elapsed_time,0)), " seconds", sep='')
    if hours == 3000:
        print("0.00001 stability not found. ",hours,"hours of survey data generated.")
    else:
        print("Stable solution found after",hours,"hours of survey data")
    print("Cars Queued = ", round(carsqueued * 100 / count_arrivals,2),"%", sep='')
    if carsqueued > 0:
        print("Average Queue time per Arrival/Queued Vehicle = ", round(queuetime / count_arrivals),"/",round(queuetime / carsqueued),"seconds")
    # print("Random Arrivals = ", round(count_arrivals / cyclecount / precision,1))
    # print("Service Time = ", servicetime)
    print("Perfect Arrivals  Demand = ", round(count_arrivals / cyclecount / precision * servicetime / 3600,2),"spaces")
    print("Random Arrivals Demand percentiles")
    for value in percentiles:
        print(value,"th - ",percentageoftime(value,count_carsparked_q)," parked and ", percentageoftime(value,count_carsqueued)," Queued", sep='')

modelrun(arrivalrate,servicetime,spaces)
